frontend/model/state.js

```diff
@@ -243,13 +243,85 @@ const getters = {
+  },
   ourIdentityContractId (state) {
     return state.loggedIn && state.loggedIn.identityContractID
   },
   // Logged In user's identity contract
   ourUserIdentityContract (state) {
     return (state.loggedIn && state[state.loggedIn.identityContractID]) || {}
   },
+  ourContributionSummary (state, getters) {
+    // TODO - Refactor Contributions and GroupPledgeGraph to use this.
+    const groupProfiles = getters.groupProfiles
+    const ourUsername = getters.ourUsername
+    const ourGroupProfile = groupProfiles[ourUsername]
+
+    if (!ourGroupProfile.incomeDetailsType) {
+      return {}
+    }
+
+    const doWeNeedIncome = ourGroupProfile.incomeDetailsType === 'incomeAmount'
+    const distribution = getters.thisMonthsPayments.frozenDistribution || getters.groupIncomeDistribution
+
+    const nonContributionsOf = (username) => groupProfiles[username].nonMonetaryContributions || []
+    const getDisplayName = (username) => getters.globalProfile(username).displayName || username
+
+    return {
+      givesMonetary: (() => {
+        if (doWeNeedIncome) { return null }
+        const who = []
+        const total = distribution
+          .filter(p => p.from === ourUsername)
+          .reduce((acc, payment) => {
+            who.push(getDisplayName(payment.to))
+            return acc + payment.amount
+          }, 0)
+
+        return { who, total, pledged: ourGroupProfile.pledgeAmount }
+      })(),
+      receivesMonetary: (() => {
+        if (!doWeNeedIncome) { return null }
+        const needed = getters.groupSettings.mincomeAmount - ourGroupProfile.incomeAmount
+        const who = []
+        const total = distribution
+          .filter(p => p.to === ourUsername)
+          .reduce((acc, payment) => {
+            who.push(getDisplayName(payment.from))
+            return acc + payment.amount
+          }, 0)
+
+        return { who, total, needed }
+      })(),
+      receivesNonMonetary: (() => {
+        const listWho = Object.keys(groupProfiles)
+          .filter(username => username !== ourUsername && nonContributionsOf(username).length > 0)
+        const listWhat = listWho.reduce((contr, username) => {
+          const displayName = getDisplayName(username)
+          const userContributions = nonContributionsOf(username)
+
+          userContributions.forEach((what) => {
+            const contributionIndex = contr.findIndex(c => c.what === what)
+            contributionIndex >= 0
+              ? contr[contributionIndex].who.push(displayName)
+              : contr.push({ who: [displayName], what })
+          })
+          return contr
+        }, [])
+
+        return listWho.length > 0 ? { what: listWhat, who: listWho } : null
+      })(),
+      givesNonMonetary: (() => {
+        const contributions = ourGroupProfile.nonMonetaryContributions
+
+        return contributions.length > 0 ? contributions : null
+      })()
+    }
```

Comment:
There are a few issues here:

1. This code belongs in ContributionsWidget.vue, not here. The getters in this file must only be functions that are used throughout the app, and they must be super efficient and compact — i.e. primitive building blocks only. This function is neither highly used throughout the app (it is specific to ContributionsWidget.vue, the use in PaymentsSummary.vue does not need it), nor is it super efficient or compact
2. This function defines and runs code for at least 4 different functions, regardless of whether it's needed or not. It could probably be done more cleanly when moved into ContributionsWidget.vue

Reply:
1. This getter can also be used at `IncomeDetailsGraph.js` and `Contributions.js`. Yes, we only need it 3 times so far, but it's kinda complex and duplicating that code doesn't seem wise.
2. It runs 4 functions, but each one of them has an early return in case it doesn't make sense to perform further logic. I don't see a problem to be as it is for now. When we reuse it at the places mentioned before, if needed, we can evolve it.

Reply:
This getter can also be used at IncomeDetailsGraph.js and Contributions.js. Yes, we only need it 3 times so far, but it's kinda complex and duplicating that code doesn't seem wise.
It runs 4 functions, but each one of them has an early return in case it doesn't make sense to perform further logic. I don't see a problem to be as it is for now. When we reuse it at the places mentioned before, if needed, we can evolve it.

Reply:
> This getter can also be used at IncomeDetailsGraph.js and Contributions.js. Yes, we only need it 3 times so far, but it's kinda complex and duplicating that code doesn't seem wise.

I can't find those files in the project... Do you mean `IncomeDetails.vue` and `Contributions.vue`?

I agree that code duplication isn't a good idea. However, if you're going to de-duplicate code, you should do so, while also making sure to do it as cleanly as possible. Neither of those files has been modified in this PR. If you're adding code that does the same thing, then those files should be modified.

That still doesn't address the issue of this function being very messy, inefficient, and bloating the `state.js` file.

There has to be a cleaner and more modular way to do this, and a way in which code is run only when it needs to be run. See if you can clean this up a bit while DRYing the other files too, and while keeping changes to `state.js` to a minimum (that file needs to be smaller than it currently is, even without these additions). You've got time since Pierre's PR needs to be merged first. If you're not able to come up with anything I can lend a hand if you'd like, but also do give it your best shot since I also need to write that blog post I've been putting off.

Reply:
Yes, I did mean .vue files, not .js.

> and a way in which code is run only when it needs to be run.
This code only runs once, when it's checked.

 I'll take a closer look on these as you suggested and let you know when it's ready.

Reply:
done @taoeffect. I refactored `Contributions.vue` to use the getter `ourContributionsSummary` and it needed all the 4 functions as I was expecting. So, in summary, this getter is used twice and 100% needed, no code could be removed.

I strongly believe that the right place for `ourContributionsSummary` is as a getter, as it is. The main advantage of using getters instead of a "standard" util function, is that a getter is only run once in the entire app, even if called multiple times (in our case, it's called twice - on the widget and on the contributions page).

> The thing is, state.js is already bigger than it should be.

Totally agree, but this isn't a problem of `ourContributionsSummary`. I mentioned some time ago that `state.js` as it is today, is not scalable. We should take better advantage of Vuex modules, and split the `state.js` into modules, [(its a common pattern for large vuex apps)](https://medium.com/3yourmind/large-scale-vuex-application-structures-651e44863e2f). Well, this is a very valid point that requires further discussion, but I don't see this directly related to this PR. If you want, I can create an issue for us to discuss state there.

Reply:
Done. https://github.com/okTurtles/group-income-simple/issues/804

Reply:
Done. #804